src/persist-client/src/read.rs

```diff
@@ -644,6 +643,11 @@ mod tests {
         let mut snapshot = read.expect_snapshot(2).await;
         let mut listen = read.expect_listen(0).await;
 
+        // Manually advance the listener's machine so that it has the latest
```

Comment:
I don't love this! some other options:
- snapshot currently clones the machine so that the methods can be `&self` instead of `&mut self` but I don't think that's super important. removing that clone would happen to make this unnecessary because the listen would inherit the state from the snapshot call
- change the `listen` call to try fetching updated state once if it's not immediately serveable
- dunno something else?

Reply:
looking right now! sorry about dropping this!

Reply:
I think snapshot mutating the reader makes the most sense to me! intuitively, having a mechanism whereby a reader changes after snapshotting seems to make sense, and it seems like we're doing

```
 // Hack: Keep this method `&self` instead of `&mut self` by cloning the
 // cached copy of the state, updating it, and throwing it away
 // afterward.
 ```
 
 purely as a means to keep that method `&self`, which i think might have the rationale that its more like the expectation for the api? I don't know of any stronger reason, and given all of that, I feel like `&mut self` is a fine way forward!

Reply:
When I was rewriting this test locally, to fit with the previous (and now unchanged!) semantics of listen, I changed this to  first do one `next()` call on listen, asserted against that. Then made it unreliable, and then fetched the rest of the listen events. Also slightly awkward, but doesn't require calling internal methods or changing signatures. :man_shrugging:

Reply:
I think it makes a lot of sense to make snapshot `&mut` and remove the machine clone (and we should consider doing that independantly), but after reverting my change to also make listen wait for as_of to be available, it seems pretty subtle for the test to rely on the fact that we call `snapshot` first. went with aljoscha's suggestion